pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/TNCudaBase.hpp

```diff
@@ -60,133 +57,194 @@ class TNCudaBase : public TensornetBase<PrecisionT, Derived> {
-    // runtime in the C++ layer
-    explicit TNCudaBase(const std::size_t numQubits, DevTag<int> dev_tag)
-        : BaseType(numQubits), handle_(make_shared_tncuda_handle()),
-          cublascaller_(make_shared_cublas_caller()), dev_tag_(dev_tag),
-          gate_cache_(std::make_shared<TNCudaGateCache<PrecisionT>>(dev_tag_)) {
-        // TODO this code block could be moved to base class and need to revisit
-        // when working on copy ctor
-        PL_ABORT_IF(numQubits < 2,
-                    "The number of qubits should be greater than 1.");
-        if constexpr (std::is_same_v<PrecisionT, double>) {
-            typeData_ = CUDA_C_64F;
-            typeCompute_ = CUTENSORNET_COMPUTE_64F;
-        } else {
-            typeData_ = CUDA_C_32F;
-            typeCompute_ = CUTENSORNET_COMPUTE_32F;
-        }
+    ~TNCudaBase() = default;
 
-        PL_CUTENSORNET_IS_SUCCESS(cutensornetCreateState(
-            /* const cutensornetHandle_t */ handle_.get(),
-            /* cutensornetStatePurity_t */ purity_,
-            /* int32_t numStateModes */
-            static_cast<int32_t>(BaseType::getNumQubits()),
-            /* const int64_t *stateModeExtents */
-            reinterpret_cast<int64_t *>(BaseType::getQubitDims().data()),
-            /* cudaDataType_t */ typeData_,
-            /*  cutensornetState_t * */ &quantumState_));
+    /**
+     * @brief Get the method of a derived class object.
+     *
+     * @return std::string
+     */
+    [[nodiscard]] auto getMethod() const -> std::string {
+        return Derived::method;
     }
 
-    ~TNCudaBase() {
-        PL_CUTENSORNET_IS_SUCCESS(cutensornetDestroyState(quantumState_));
+    /**
+     * @brief Access the CublasCaller the object is using.
+     *
+     * @return a reference to the object's CublasCaller object.
+     */
+    auto getCublasCaller() const -> const CublasCaller & {
+        return *cublascaller_;
     }
 
     /**
-     * @brief Get the CUDA data type.
+     * @brief Get the full state vector representation of a quantum state.
+     * Note that users/developers should be responsible to ensure that there is
+     * sufficient memory on the host to store the full state vector.
      *
-     * @return cudaDataType_t
+     * @param res Pointer to the host memory to store the full state vector
+     * @param res_length Length of the result vector
      */
-    [[nodiscard]] auto getCudaDataType() const -> cudaDataType_t {
-        return typeData_;
+    void getData(ComplexT *res, const std::size_t res_length) {
+        PL_ABORT_IF(log2(res_length) != BaseType::getNumQubits(),
+                    "The size of the result vector should be equal to the "
+                    "dimension of the quantum state.");
+
+        std::size_t avail_gpu_memory = getFreeMemorySize();
+
+        PL_ABORT_IF(log2(avail_gpu_memory) < BaseType::getNumQubits(),
+                    "State tensor size exceeds the available GPU memory!");
```

Comment:
is this necessary because it will need to create a copy of the statevector on GPU first?

Reply:
isn't `avail_gpu_memory` in units of bytes? do we need to take care of the number of bytes per complex number for each ComplexT?

Reply:
Nice catch!

Reply:
> is this necessary because it will need to create a copy of the statevector on GPU first?
yea, the device memory will be allocated the `get_state_tensor` method.